### PR TITLE
Fix ShallowClone soundness and introduce ConstValue

### LINKED ISSUE #645: ShallowClone trait is broken
The `ShallowClone` trait on a struct is supposed to signify that cloning the struct will never yield new copies of any members that can be mutated at a later stage. This contract is currently broken in the way `ShallowClone` is used. 

For example, `LmdbImpl` implements `ShallowClone`, but it's allowed to put variables inside of it that can be mutated and Cloned. If we add a `u32` or a `bool`, nothing will break (a successful compile will happen). This basically means that cloning that struct will cause undefined behavior, because the "shallow" clone is not a shallow clone anymore, but it holds an independent value that can be mutated independently.

`ShallowClone` should be some kind of auto trait that enforces `ShallowClone` on all members of a struct. It should also cause an error if any type underneath supports deep Cloning.

### PR DESCRIPTION
**Pull request by TheQuantumPhysicist**
Fixes #645

### CODE REVIEW

File: utils/src/shallow_clone.rs [hunk 1]
```diff
@@ -17,6 +17,12 @@
 
 use std::marker::PhantomData;
 
+#[cfg(not(loom))]
+use std::sync::Arc;
+
+#[cfg(loom)]
+use loom::sync::Arc;
```
Comment (iljakuklic) [inline utils/src/shallow_clone.rs:24]:
```suggestion
use crate::sync::Arc;
```

File: utils/src/const_value.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.value.hash(state)
+    }
+}
+
+impl<T> AsRef<T> for ConstValue<T> {
+    fn as_ref(&self) -> &T {
+        &self.value
+    }
+}
```
Comment (iljakuklic) [inline utils/src/const_value.rs:100]: Also implement `Borrow`?

File: utils/src/const_value.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    fmt::{Debug, Display},
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+/// A wrapper for a type that is not supposed to be modified, no matter one
```
Comment (iljakuklic) [inline utils/src/const_value.rs:22]: Unfinished sentence?

File: utils/src/const_value.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+/// This particularly solves the problem of not being able to mark member variables as const in Rust
+pub struct ConstValue<T> {
+    value: T,
+}
+
+impl<T> ConstValue<T> {
+    pub fn new(value: T) -> Self {
+        Self { value }
+    }
+}
```
Comment (iljakuklic) [inline utils/src/const_value.rs:32]: Should there also be a method to take the inner value out of the `ConstValue` wrapper?

File: utils/src/const_value.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+use std::{
+    fmt::{Debug, Display},
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+/// A wrapper for a type that is not supposed to be modified, no matter one
+/// This particularly solves the problem of not being able to mark member variables as const in Rust
+pub struct ConstValue<T> {
```
Comment (iljakuklic) [inline utils/src/const_value.rs:24]: I understand this wrapper works on best effort basis. It blocks calling `&mut self` methods on the inner value.

As far as I can see, one can still do something like:
```rust
fn modify_const_value(cv: &mut ConstValue<u32>) {
    *cv = ConstValue::new(42);
}
```

Same applies if you're given a `&mut` of a struct containing a `ConstValue`. Also anything with interior mutability can still be modified.

File: utils/src/shallow_clone.rs [hunk 6]
```diff
@@ -15,20 +15,67 @@
...
+    }
+}
+
+impl<T> ShallowClone for PhantomData<T> {
+    fn shallow_clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T: Clone> ShallowClone for ConstValue<T> {
```
Comment (iljakuklic) [inline utils/src/shallow_clone.rs:74]: A comment explaining why it's okay to `shallow_clone` a `ConstValue` would be nice

**Review by iljakuklic — APPROVED** [review]
Looks good. Some minor suggestions and comments below

File: utils/src/const_value.rs (hunk 4 above)
Comment (TheQuantumPhysicist) [reply utils/src/const_value.rs:32]: Done.

File: utils/src/const_value.rs (hunk 2 above)
Comment (TheQuantumPhysicist) [reply utils/src/const_value.rs:100]: Done

**Review by iljakuklic — APPROVED** [review]

**Pull request merged**